provider/pkg/resources/cfn_custom_resource.go

```diff
@@ -0,0 +1,681 @@
+	})
+	updateDuration := c.clock.Since(startTime)
+	glog.V(9).Infof("%s custom resource update took %s", label, updateDuration)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to update custom resource: %w", err)
+	}
+
+	if response.Status == cfn.StatusFailed {
+		glog.V(9).Infof("%s custom resource update failed: %s", label, response.Reason)
+		return nil, fmt.Errorf("failed to update custom resource: %s", response.Reason)
+	}
+
+	glog.V(9).Infof("%s custom resource update succeeded", label)
+
+	// if the physical resource ID has changed, we need to delete the old resource
+	if response.PhysicalResourceID != typedState.PhysicalResourceID {
+		glog.V(9).Infof("%s physical resource ID changed from %q to %q, deleting old resource", label, typedState.PhysicalResourceID, response.PhysicalResourceID)
+
+		deleteEvent := cfn.Event{
+			PhysicalResourceID: typedState.PhysicalResourceID,
+			RequestType:        cfn.RequestDelete,
+			ResourceType:       typedState.ResourceType,
+			LogicalResourceID:  urn.Name(),
+			StackID:            typedState.StackID,
+			ResourceProperties: oldTypedInputs.CustomResourceProperties,
+		}
+
+		deleteTimeout := DefaultCustomResourceTimeout
+		// if a custom timeout is set, the delete operation is allowed to take the remaining time
+		// otherwise we allow it to take the default timeout
+		if timeout != 0 {
+			deleteTimeout = timeout - updateDuration
+			glog.V(9).Infof("%s custom resource update took %s, clean up of the old resource is allowed to take %s", label, updateDuration, deleteTimeout)
+			if deleteTimeout <= 0 {
+				return nil, fmt.Errorf("failed to clean up old custom resource: not enough time left to delete the old resource. Consider increasing the timeout")
+			}
+		}
+
+		deleteResponse, err := c.invokeCustomResource(ctx, customResourceInvokeData{
+			event:           deleteEvent,
+			bucket:          newTypedInputs.BucketName,
+			bucketKeyPrefix: newTypedInputs.BucketKeyPrefix,
+			timeout:         deleteTimeout,
+			loggingLabel:    label,
+			serviceToken:    newTypedInputs.ServiceToken,
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to clean up old custom resource: %w", err)
+		}
+		if deleteResponse.Status == cfn.StatusFailed {
+			return nil, fmt.Errorf("failed to clean up old custom resource %q: %s", typedState.PhysicalResourceID, deleteResponse.Reason)
+		}
+		glog.V(9).Infof("%s old custom resource %q successfully cleaned up", label, typedState.PhysicalResourceID)
+	}
+
+	outputs := newTypedInputs.makeOutputs(inputs, response)
+	return outputs, nil
+}
+
+// Read returns the current inputs and outputs of the custom resource because CFN custom resources do not store state.
+// They are just a stateless wrapper around a Lambda function or SNS topic.
+func (c *cfnCustomResource) Read(ctx context.Context, urn urn.URN, id string, oldInputs resource.PropertyMap, oldState resource.PropertyMap) (resource.PropertyMap, resource.PropertyMap, bool, error) {
+	if len(oldState) == 0 {
```

Comment:
Comment here that len(oldState) is trying to guess if we are in a Read-for-import. 
The developer docs are improving but not quite there yet but are being improved (cc @lunaris )

https://pulumi-developer-docs.readthedocs.io/latest/developer-docs/providers/implementers-guide.html#read